local baseUrl = "https://raw.githubusercontent.com/nowkyyl/AutoCatch/refs/heads/main/src"
local function require(path, raw): string
    path = path:gsub("^@self/", "")
    local response = request({
        Url = `{baseUrl}/{path}.luau` :: string,
        Method = "GET"
    })

    assert(response.Success, `Failed to load {path}.luau`)
    if not raw then
        local chunk = assert(loadstring(response.Body))
        return chunk()
    end
    return response.Body
end

local construct = loadstring(require("init", true))
setfenv(construct, (setmetatable({
    require = require
}, { __index = getfenv(construct) }) :: any))
construct()
